Use containerd as source of truth

carlosedp opened the pull request:
## Description

This allow persisting state between restarts by reading containers from
containerd on certain task status.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [ ] I have raised an issue to propose this change **this is required**

Currently faas-containerd does not persist data between restarts so deployed functions cannot be used if faas-containerd process is restarted.

Addresses issue #20 .

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Verified the basic service functions to Update, Add, Get and List the containers. Still work-in-progress.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Commits:

- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] My commit message has a body and describe how this was tested and why it is required.
- [x] I have signed-off my commits with `git commit -s` for the Developer Certificate of Origin (DCO)

Code:

- [x] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.

Docs:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

alexellis commented on handlers/deploy.go:
```diff
@@ -155,6 +155,7 @@ func deploy(ctx context.Context, req types.FunctionDeployment, client *container
 	ip := &result.Interfaces[defaultInterface].IPConfigs[0].IP
 
 	serviceMap.Add(name, ip)
+	serviceMap.Update()
```
What does Update do?

Add/Update didn't come across intuitive

alexellis commented on handlers/read.go:
```diff
@@ -14,10 +14,19 @@ func MakeReadHandler(client *containerd.Client, serviceMap *ServiceMap) func(w h
 
 		res := []types.FunctionStatus{}
 		for _, k := range serviceMap.Keys() {
+			i, _ := serviceMap.containers[k].Image(serviceMap.context)
```
serviceMap.context? is that just a context.Context? Why not create one here for use?

alexellis commented on handlers/read.go:
```diff
@@ -14,10 +14,19 @@ func MakeReadHandler(client *containerd.Client, serviceMap *ServiceMap) func(w h
+				Name:      k,
+				Image:     i.Name(),
+				Replicas:  uint64(replicas),
+				Namespace: "openfaas-fn",
```
eventually I want to extract this variable so that it can be changed.

carlosedp replied:
I'll refactor all this and put everything that's network related in the same file.

alexellis replied:
No need to do that now. Best control the amount of changes.

alexellis commented on handlers/service_map.go:
```diff
@@ -1,20 +1,55 @@
+	}
+}
+
+func (s *ServiceMap) Update() {
```
Maybe add a comment and change Update to something else, what's it meant to do vs Add?

alexellis commented on handlers/service_map.go:
```diff
@@ -1,20 +1,55 @@
+	containers, _ := s.client.Containers(s.context)
+	for _, k := range containers {
+		i := k.ID()
+		s.containers[i] = k
```
Should we really capture a handle to the containerd object, or have some safe DTO struct to keep the key info we want?

alexellis commented on handlers/service_map.go:
```diff
@@ -49,3 +84,19 @@ func (s *ServiceMap) Delete(name string) {
+	if addrsErr != nil {
+		return nil, fmt.Errorf("Unable to find address for veth pair using: %v %s", peerIDs, addrsErr)
+	}
+	return &addrs[0].CIDRs[0].IP, nil
```
Check the length before returning, otherwise return error

alexellis reviewed (commented):
Looks good, as expected, just a few comments on style.

alexellis commented:
> How Has This Been Tested?
> Verified the basic service functions to Update, Add, Get and List the containers. Still work-in-progress.

What about restarting faas-containerd (tasks stay), or restarting the host (tasks die)?

Have you tried a scale to zero? replicas should be 0 if not running, so the scale to zero pauses them.. and you should see them unpause when invoking through faasd and the gateway.

alexellis commented:
I checked out the PR, built it and deployed with CNI 0.8.4 on x86_64, this is what I got

```
Jan 10 18:39:31 alexx systemd[1]: faas-containerd.service: Unit entered failed state.
Jan 10 18:39:31 alexx systemd[1]: faas-containerd.service: Failed with result 'exit-code'.
Jan 10 18:39:38 alexx systemd[1]: Stopped faasd-containerd.
Jan 10 18:39:38 alexx systemd[1]: Started faasd-containerd.
Jan 10 18:39:38 alexx faas-containerd[7346]: 2020/01/10 18:39:38 faas-containerd starting..        Version:         Commit:         Service Timeout: 1m0s
Jan 10 18:39:38 alexx faas-containerd[7346]: panic: runtime error: invalid memory address or nil pointer dereference
Jan 10 18:39:38 alexx faas-containerd[7346]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x90 pc=0xf174d5]
Jan 10 18:39:38 alexx faas-containerd[7346]: goroutine 1 [running]:
Jan 10 18:39:38 alexx faas-containerd[7346]: github.com/alexellis/faas-containerd/handlers.(*ServiceMap).Update(0xc00028a6e0)
Jan 10 18:39:38 alexx faas-containerd[7346]:         /home/alex/go/src/github.com/alexellis/faas-containerd/handlers/service_map.go:46 +0x1b5
Jan 10 18:39:38 alexx faas-containerd[7346]: main.Start()
Jan 10 18:39:38 alexx faas-containerd[7346]:         /home/alex/go/src/github.com/alexellis/faas-containerd/main.go:105 +0x6e8
Jan 10 18:39:38 alexx faas-containerd[7346]: main.main()
Jan 10 18:39:38 alexx faas-containerd[7346]:         /home/alex/go/src/github.com/alexellis/faas-containerd/main.go:51 +0x20
Jan 10 18:39:38 alexx systemd[1]: faas-containerd.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jan 10 18:39:38 alexx systemd[1]: faas-containerd.service: Unit entered failed state.
Jan 10 18:39:38 alexx systemd[1]: faas-containerd.service: Failed with result 'exit-code'.
```

alexellis commented:
This is caused due to a lack of proper error handling, if a function returns two things - a result and an error, then you cannot use that result if there was an error.


I'd suggest the following:

```golang
func (s *ServiceMap) Update() {
	s.lock.Lock()
	defer s.lock.Unlock()

	containers, _ := s.client.Containers(s.context)
	for _, k := range containers {
		id := k.ID()
		s.containers[id] = k

		t, err := k.Task(s.context, nil)

		if err != nil {
			log.Printf("No task for %s\n", id)
		} else {
			svc, statusErr := t.Status(s.context)

			if statusErr != nil && svc.Status == "running" {
				s.servicesPID[id] = t.Pid()
				// Get container IP address
				ip, ipErr := getIP(DefaultBridgeName, int(t.Pid()))
				if ipErr != nil {
					s.services[id] = ip
				}
			}
		}
	}
}
```

If t doesn't exist, then update replicas to 0.

alexellis commented:
After fixing that, I get:

```
faas-cli store deploy "ASCII Cows" --name cows1
WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.

Unexpected status: 502, message: 

Function 'cows1' failed to deploy with status code: 502

```

faasd after merging https://github.com/alexellis/faasd/pull/21 shows:

```
Jan 10 18:48:09 alexx faasd[13600]: 2020/01/10 18:48:09 Get http://faas-containerd:8081/system/functions: dial tcp 10.62.0.1:8081: connect: connection refused
```

alexellis commented:
```
diff --git a/handlers/service_map.go b/handlers/service_map.go
index 4e517b1..c07273b 100644
--- a/handlers/service_map.go
+++ b/handlers/service_map.go
@@ -3,6 +3,7 @@ package handlers
 import (
 	"context"
 	"fmt"
+	"log"
 	"net"
 	"sync"
 
@@ -40,15 +41,24 @@ func (s *ServiceMap) Update() {
 
 	containers, _ := s.client.Containers(s.context)
 	for _, k := range containers {
-		i := k.ID()
-		s.containers[i] = k
-		t, _ := k.Task(s.context, nil)
-		svc, _ := t.Status(s.context)
-		if svc.Status == "running" {
-			s.servicesPID[i] = t.Pid()
-			// Get container IP address
-			ip, _ := getIP(DefaultBridgeName, int(t.Pid()))
-			s.services[i] = ip
+		id := k.ID()
+		s.containers[id] = k
+
+		t, err := k.Task(s.context, nil)
+
+		if err != nil {
+			log.Printf("No task for %s\n", id)
+		} else {
+			svc, statusErr := t.Status(s.context)
+
+			if statusErr != nil && svc.Status == "running" {
+				s.servicesPID[id] = t.Pid()
+				// Get container IP address
+				ip, ipErr := getIP(DefaultBridgeName, int(t.Pid()))
+				if ipErr != nil {
+					s.services[id] = ip
+				}
+			}
 		}
 	}
 }
diff --git a/main.go b/main.go
index 874479e..174f39c 100644
--- a/main.go
+++ b/main.go
@@ -99,6 +99,7 @@ func Start() {
 	if err != nil {
 		panic(err)
 	}
+
 	defer client.Close()
 
 	serviceMap := handlers.NewServiceMap(client, "openfaas-fn")

```

Feel free to apply this patch

alexellis commented:
My original suggestion was to remove the service map entirely and go directly to containerd all the time, then once we would have had that stable, I would have re-written the cache and optimisations.

alexellis commented:
@carlosedp please can you reply to the comments I've left for you, and see if you have addressed or have any questions?

alexellis commented:
/set title: Use containerd as source of truth

alexellis commented:
@carlosedp this will need a rebase now due to the merge of #26

carlosedp commented:
Working on this already rebased on #26

alexellis commented on handlers/read.go:
```diff
@@ -14,10 +14,19 @@ func MakeReadHandler(client *containerd.Client, serviceMap *ServiceMap) func(w h
+			t, _ := serviceMap.containers[k].Task(serviceMap.context, nil)
+			svc, _ := t.Status(serviceMap.context)
+			replicas := 0
+			if svc.Status == "running" {
```
:muscle:

alexellis commented:
I'm not seeing any changes for `handlers/replicas.go`, did you cover it?

carlosedp committed:
Keep containerd as the source of truth

Removed ServiceMap struct and changed all operations to be performed on
containerd.
All status for the running and stopped tasks are read by the current
operation being executed.

Signed-off-by: Carlos de Paula <me@carlosedp.com>

carlosedp commented:
Force-pushed new commit where all ServiceMap struct has been removed and all state comes from containerd.

alexellis commented on handlers/cni_network.go:
```diff
@@ -109,11 +109,13 @@ func DeleteCNINetwork(ctx context.Context, cni gocni.CNI, client *containerd.Cli
-		log.Printf("[Delete] removing CNI network for %s\n", task.ID())
 		if err != nil {
-			return errors.Wrapf(err, "Unable to get task for container %s: %s", name, err)
+			log.Printf("[Delete] container %s does not have task\n", name)
```
Should probably say `a task`, but how about: `Unable to find a task for container: %s\n`

alexellis commented on handlers/containers.go:
```diff
@@ -0,0 +1,71 @@
+			image:     image.Name(),
+		}
+		replicas := 0
+		task, err := c.Task(ctx, nil)
```
If the task is in a frozen state, the replicas should also be 0, this could be fixed in a follow-up PR

alexellis commented on handlers/containers.go:
```diff
@@ -0,0 +1,71 @@
+			// Task for container exists
+			svc, err := task.Status(ctx)
+			if err != nil {
+				return Function{}, fmt.Errorf("Unable to get task status for container: %s", name, err)
```
In Go, errors should start with lowercase

alexellis replied:
Small bug here too, 2 args, but only one `%s`

alexellis commented on handlers/invoke_resolver.go:
```diff
@@ -4,25 +4,26 @@ import (
 
-	serviceIP := i.serviceMap.Get(functionName)
-	if serviceIP == nil {
+	fun, err := GetFunction(i.client, functionName)
```
can you write `function` or `fn`, vs `fun`? Just a nit

alexellis reviewed (commented):
Looks very good, just a couple of nits / questions.

Can you also confirm what paths have been tested in the latest push/change?

alexellis commented:
I'm going to merge and ask for you to work out anything else that needs a change in HEAD / master.

alexellis commented on handlers/containers.go:
```diff
@@ -0,0 +1,71 @@
+		return f, nil
+
+	}
+	return Function{}, fmt.Errorf("Unable to find function %s: %s", name, err)
```
Possible typo, `%s:` should normally be:

`function: NAME`
